tests/test_constraint_verifier.py

```diff
@@ -0,0 +1,234 @@
+            .given(e1, e2, e3) \
+            .penalizes(2)
+
+    with pytest.raises(AssertionError):
+        verifier.verify_that(same_value) \
+            .given(e1, e2, e3) \
+            .penalizes(4)
+
+    verifier.verify_that(same_value) \
+        .given_solution(solution) \
+        .penalizes(3)
+
+    verifier.verify_that(same_value) \
+        .given_solution(solution) \
+        .penalizes()
+
+    with pytest.raises(AssertionError):
+        verifier.verify_that(same_value) \
+            .given_solution(solution) \
+            .rewards(3)
+
+    with pytest.raises(AssertionError):
+        verifier.verify_that(same_value) \
+            .given_solution(solution) \
+            .penalizes(2)
+
+    with pytest.raises(AssertionError):
+        verifier.verify_that(same_value) \
+            .given_solution(solution) \
+            .penalizes(4)
+
+    verifier.verify_that(is_value_one) \
+        .given(e1, e2, e3) \
+        .rewards(3)
+
+    verifier.verify_that(is_value_one) \
+        .given(e1, e2, e3) \
+        .rewards()
+
+    with pytest.raises(AssertionError):
+        verifier.verify_that(is_value_one) \
+            .given(e1, e2, e3) \
+            .penalizes()
+
+    with pytest.raises(AssertionError):
+        verifier.verify_that(is_value_one) \
+            .given(e1, e2, e3) \
+            .penalizes(3)
+
+    with pytest.raises(AssertionError):
+        verifier.verify_that(is_value_one) \
+            .given(e1, e2, e3) \
+            .rewards(2)
+
+    with pytest.raises(AssertionError):
+        verifier.verify_that(is_value_one) \
+            .given(e1, e2, e3) \
+            .rewards(4)
+
+
+def test_constraint_verifier_create():
+    @optapy.problem_fact
+    class Value:
+        def __init__(self, code):
+            self.code = code
+
+        def __str__(self):
+            return f'Value(code={self.code})'
+
+
+    @optapy.planning_entity
+    class Entity:
+        def __init__(self, code, value=None):
+            self.code = code
+            self.value = value
+
+        @optapy.planning_variable(Value, value_range_provider_refs=['value_range'])
+        def get_value(self):
+            return self.value
+
+        def set_value(self, value):
+            self.value = value
+
+
+    def same_value(constraint_factory: optapy.constraint.ConstraintFactory):
+        return (constraint_factory.for_each(Entity)
+                    .join(Entity, optapy.constraint.Joiners.less_than(lambda e: e.code),
+                                  optapy.constraint.Joiners.equal(lambda e: e.value))
+                    .penalize('Same value', optapy.score.SimpleScore.ONE)
+                )
+
+    def is_value_one(constraint_factory: optapy.constraint.ConstraintFactory):
+        return (constraint_factory.for_each(Entity)
+                    .filter(lambda e: e.value.code == 'v1')
+                    .reward('Value 1', optapy.score.SimpleScore.ONE)
+                )
+
+    @optapy.constraint_provider
```

Comment:
Another annotation that I don't think has an equivalent in OptaPlanner.

Reply:
It does; this is the `ConstraintProvider` interface, which is a functional interface (has only one method, `defineConstraints`).